keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+  process, it must also be usable in local development to generate one-off
+  packages.
+
+- <a id="pkg-list"></a>The tooling must support packaging
+  - `kubelet`
+  - `kubectl`
+  - `kubeadm`
+  - `cri-tools`
+  - `kubernetes-cni`
+
+  It should be easy to integrate new packages, especially when they are
+  one-binary packages
+
+- <a id="plat-list"></a>The tooling must support packaging for multiple platforms
+  - `linux/amd64`
+  - `linux/arm64`
+  - `linux/armhf`
+  - `linux/ppc64el`
+  - `linux/s390x`
+
+  It should be easy to integrate other platforms, given there is general
+  support by the supported distributions and the cross-building tools.
+  Examples: `kubectl` for `darwin/amd64`, `kubelet` for `windows/386`, ...
+
+- The tooling must support building `deb`s and `rpm`s
+
+  It should be possible to add new package types, e.g. NuGet for Windows
+
+- The tooling must support signing the generated packages
+
+- The documentation for the process of building packages, local one-off builds
+  and as part of the release process, must be excellent.
+
+- Any other option to build packages must be retired eventually
+
+  There should be documentation how and where the old packaging option can be
+  found and how they can be used for cases where someone wants to build
+  packages before the new tools were introduced.
+
+- It must be equally easy to build packages from
+  - the currently checked out revision (any revision after the tooling has been
+    introduced into `k/k`)
+  - a pre-build binary (pulling from some well-defined staging bucket or that like)
+
+- It must be containerized and must not depend on anything more then docker on
+  the host
+
+### Non-Goals
+
+- Uploading, publishing, ... of packages
+- Clean up [`k/k/build/`][k-k] (remove outdated scripts & documentation) other
+  then "competitor" packaging mechanism
+
+## Proposal
+
+All the following steps are only considered done when there is proper documentation on
+- how to use those tools
+- how they interact with each other
+- how to (re-)build them
+- how to update them
+- how their configuration works
+
+**Note**: This section refers to a couple of specifics that might be seen as
+implementation details in here (e.g. references to a specific tools or specific
+filesystem layouts). Those are just there to illustrate in more detail what we
+aim towards, the details and actual tools can absolutely change in the actual
+implementation. 
+
+- [ ] Define and document the package configuration
+
+  The data referred to in this KEP as package configuration might be something
+  like <a id="nfpm-config">a `nfpm` (compatible) configuration file</a> plus a set
+  of additional files referred to in the [main configuration
+  file](#nfpm-config), potentially organised in subdirectories
+
+  ```sh
+  ./main.conf
+  ./linux/50-kubeadm.conf
+  ./linux/amd64/50-kubeadm.conf
+  [...]
+  ```
+
+- [ ] Create a tool that can create **one** package configuration
```

Comment:
are we creating a tool, or wrapping nfpm?
below it says:
> *Risk*: `nfpm` is not the right tool, it goes away, works differently, ...

I'm not sure if Kubernetes should be inventing & maintaining a packaging format ...

If this tool is to be a thin layer over another tool, we should call that out here, otherwise a significant risk seems to be the technical debt of maintaining our own packaging tool.

Reply:
> However I think the actual tool is an implementation detail which does not necessarily be decided on in the KEP.

i might need a discussion at least. ideally during a SIG release meeting.

Reply:
I was going for using their format in combination of a directory tree holding additional files like systemd configs -- a directory tree so that we can have different systemd files for different platforms and so.

> to me the tool is the whole detail as far as risks, because that's what we now have to maintain. we have two sets of tools today that more or less work, what we need is something converged, maintinable, with specs versioned with kubernetes imho.

 I am not sure if I understand you correctly:
- you are cautious to bring in a new tool (`nfpm` and the "orchestration" around that) because of additional maintenance burden
- you'd opt for settling for one already existing tooling (`bazel` or `k/release/{deb/rpm}`)
- you want to have the spec (and other stuff like systemd files) in `k/k`
- you have no preference if the tools themselves are in `k/k` (and thus versioned and branched like `k/k` or keeping it in `k/release`

Did I paraphrase you correctly here, @BenTheElder ?

Reply:
Most of those points yes, but as far as using a new tool vs an existing one or one of the two setups we already mostly use (bazel, k/release scripts), I don't know which makes sense currently and I think this project should look into that.

I just want to explore the options and think that we should highlight the risk of maintaining our own kubernetes project specific format + tool if we go that route, but not that we necessarily shouldn't. Depending on tools that aren't effective enough is also a risk :+1: 

One tool would be great.

My own personal first priority is definitely:
- We really need to get the specs like the systemd files into kubernetes/kubernetes canonically for any and all tools that we use, so there is one source of truth and it is versioned with the binaries they belong to

Tooling can live either place. we do already have a lot of build tooling in kubernetes, so maybe that's the right place, but I always hesitate about packing more things into the core repo, EG in theory tools like kubeadm may move out one day, maybe the build tool should be in a common repo.

I won't push hard on that at all either, I just wanted to bring those up for consideration. PRs to kubernetes/kubernetes can be excrutiating :grimacing: